README.md
mixpanel-py3
============
This library allows for server-side integration of Mixpanel.
This is an in-official Mixpanel library for Python3.
Its major version number reflects the Mixpanel API version.

Features over the official client
---------------------------------
* Support for Python3
* Optional async consumer using tornado's AsyncHTTPClient
* Convenient method to generate tracking URLs, useful for email tracking

Installation
------------
This library can be installed via PyPI:

    pip install mixpanel-py3

Getting Started
---------------
Typical usage usually looks like this:

    from mixpanel import Mixpanel

    mp = Mixpanel(YOUR_TOKEN)

    # tracks an event with certain properties
    mp.track('button clicked', {'color': 'green', 'size': 'large'})

    # sends an update to a user profile
    mp.people_set(USER_ID, {'$first_name': 'Graham', 'favorite color': 'red'})

You can use an instance of the Mixpanel class for sending all of your events and people updates.

Issues
------
If you find any issues please file those on [GitHub](https://github.com/MyGGaN/mixpanel-python/issues) preferably with a pull request.

Contribute
----------
If you have improved code, tests, documentation or added a feature, please open a pull request and share your work.

Additional Information
----------------------
[mixpanel-py](https://github.com/mixpanel/mixpanel-python) Official client for Python2.

[Tornado](http://www.tornadoweb.org/en/stable/) I/O-loop based web framework.


CHANGES.txt
v3.0.4
* Fixed broken test case for get_tracking_url()
* Making alias() synchronous by default

v3.0.3
* Add get_tracking_url()

v3.0.2
* Add dummy flush() to Consumer, fixes #1

v3.0.0
* Initial release based on mixpanel-py v3.1.2


test_async.py
__author__ = 'Fredrik Svensson'

# TODO:
# - Add negative test cases
# - Add tests with unicode chars
# - Test so Exceptions are raised

import contextlib
from unittest.mock import MagicMock, patch
import mixpanel
from mixpanel.async_consumer import AsyncConsumer, BufferedAsyncConsumer
import unittest
import urllib.parse
import base64
import json

try:
    import tornado.testing
    from tornado.concurrent import TracebackFuture
    from tornado import stack_context
except ImportError:
    tornado_loaded = False
else:
    tornado_loaded = True


@unittest.skipUnless(tornado_loaded, "tornado module not loaded")
class AsyncTest(tornado.testing.AsyncTestCase):
    def mock_fetch(self, request, callback=None, **kwargs):
        future = TracebackFuture()
        if callback is not None:
            callback = stack_context.wrap(callback)

            def handle_future(future):
                response = future.result()
                self.io_loop.add_callback(callback, response)
            future.add_done_callback(handle_future)

        res = MagicMock()
        future.set_result(res)

        return future

class ConsumerAsyncTestCase(AsyncTest):
    def setUp(self):
        super(ConsumerAsyncTestCase, self).setUp()
        self.consumer = AsyncConsumer()

    @contextlib.contextmanager
    def _assertSends(self, expect_url, expect_data):

        with patch('tornado.httpclient.AsyncHTTPClient.fetch',
                   return_value=self.mock_fetch) as http_client:
            yield

            self.assertEqual(http_client.call_count, 1)
            ((url, _), req) = http_client.call_args
            self.assertEqual(url, expect_url)
            body = req['body'].decode('utf-8')
            self.assertEqual(set(body.split('&')),
                             set(expect_data.split('&')))

    @tornado.testing.gen_test
    def test_send_events(self):
        with self._assertSends('https://api.mixpanel.com/track', 'ip=0&data=IkV2ZW50Ig%3D%3D&verbose=1'):
            self.consumer.send('events', '"Event"')

    @tornado.testing.gen_test
    def test_send_people(self):
        with self._assertSends('https://api.mixpanel.com/engage','ip=0&data=IlBlb3BsZSI%3D&verbose=1'):
            self.consumer.send('people', '"People"')


class FunctionalAsyncTestCase(AsyncTest):
    def setUp(self):
        super(FunctionalAsyncTestCase, self).setUp()
        self.TOKEN = '12345'
        self.mp = mixpanel.Mixpanel(self.TOKEN, consumer=AsyncConsumer())
        self.mp._now = lambda : 1000

    @contextlib.contextmanager
    def _assertRequested(self, expect_url, expect_data):
        with patch('tornado.httpclient.AsyncHTTPClient.fetch',
                   return_value=self.mock_fetch) as http_client:
            yield

            self.assertEqual(http_client.call_count, 1)
            ((url, _), req) = http_client.call_args
            self.assertEqual(url, expect_url)
            body = req['body'].decode('utf-8')
            data = urllib.parse.parse_qs(req['body'].decode('utf-8'))
            self.assertEqual(len(data['data']), 1)
            payload_encoded = data['data'][0]
            payload_json = base64.b64decode(payload_encoded).decode('utf-8')
            payload = json.loads(payload_json)
            self.assertEqual(payload, expect_data)

    @tornado.testing.gen_test
    def test_track_functional(self):
        expect_data = {'event': {'color': 'blue', 'size': 'big'}, 'properties': {'mp_lib': 'python', 'token': self.TOKEN, 'distinct_id': 'button press', '$lib_version': mixpanel.VERSION, 'time': 1000}}
        with self._assertRequested('https://api.mixpanel.com/track', expect_data):
            self.mp.track('button press', {'size': 'big', 'color': 'blue'})


    @tornado.testing.gen_test
    def test_people_set_functional(self):
        expect_data = {'$distinct_id': 'amq', '$set': {'birth month': 'october', 'favorite color': 'purple'}, '$time': 1000000, '$token': self.TOKEN}
        with self._assertRequested('https://api.mixpanel.com/engage', expect_data):
             self.mp.people_set('amq', {'birth month': 'october', 'favorite color': 'purple'})


class BufferedConsumerAsyncTestCase(AsyncTest):
    def setUp(self):
        super(BufferedConsumerAsyncTestCase, self).setUp()
        self.MAX_LENGTH = 10
        self.consumer = BufferedAsyncConsumer(self.MAX_LENGTH)

    def test_buffer_hold_and_flush(self):
        with patch('tornado.httpclient.AsyncHTTPClient.fetch',
                   return_value=self.mock_fetch) as http_client:
            self.consumer.send('events', '"Event"')
            self.assertEqual(http_client.call_count, 0)

            self.consumer.flush()

            self.assertEqual(http_client.call_count, 1)
            ((url, _), req) = http_client.call_args
            self.assertEqual(url, 'https://api.mixpanel.com/track')
            body = req['body'].decode('utf-8')
            self.assertEqual(set(body.split('&')),
                             set('ip=0&data=WyJFdmVudCJd&verbose=1'.split('&')))

    def test_buffer_fills_up(self):
        with patch('tornado.httpclient.AsyncHTTPClient.fetch',
                   return_value=self.mock_fetch) as http_client:
            for i in range(self.MAX_LENGTH - 1):
                self.consumer.send('events', '"Event"')
            self.assertEqual(http_client.call_count, 0)

            self.consumer.send('events', '"Last Event"')

            self.assertEqual(http_client.call_count, 1)
            ((url, _), req) = http_client.call_args
            self.assertEqual(url, 'https://api.mixpanel.com/track')
            body = req['body'].decode('utf-8')
            self.assertEqual(set(body.split('&')),
                             set('ip=0&data=WyJFdmVudCIsIkV2ZW50IiwiRXZlbnQiLCJFdmVudCIsIkV2ZW50IiwiRXZlbnQiLCJFdmVudCIsIkV2ZW50IiwiRXZlbnQiLCJMYXN0IEV2ZW50Il0%3D&verbose=1'.split('&')))


if __name__ == "__main__":
    unittest.main()


tests/__init__.py
__author__ = 'Fredrik Svensson'


demo/subprocess_consumer.py

import multiprocessing
import random

from mixpanel import Mixpanel, BufferedConsumer


# As your application scales, it's likely you'll want to
# to detect events in one place and send them somewhere
# else. For example, you might write the events to a queue
# to be consumed by another process.
#
# This demo shows how you might do things, using
# a custom Consumer to consume events, and a
# and a BufferedConsumer to send them to Mixpanel
#
#
# You can provide custom communication behaviors
# by providing your own consumer object to the
# Mixpanel constructor. Consumers are expected to
# have a single method, 'send', that takes an
# endpoint and a json message.


class QueueWriteConsumer(object):
    def __init__(self, queue):
        self.queue = queue

    def send(self, endpoint, json_message):
        self.queue.put((endpoint, json_message))


def do_tracking(project_token, distinct_id, queue):
    """
    This process represents the work process where events
    and updates are generated. This might be the service
    thread of a web service, or some other process that
    is mostly concerned with getting time-sensitive work
    done.
    """
    consumer = QueueWriteConsumer(queue)
    mp = Mixpanel(project_token, consumer)
    for i in range(100):
        event = 'Tick'
        mp.track(distinct_id, event, {'Tick Number': i})
        print('tick {0}'.format(i))

    queue.put(None)  # tell worker we're out of jobs


def do_sending(queue):
    """
    This process is the analytics worker process- it can
    wait on HTTP responses to Mixpanel without blocking
    other jobs. This might be a queue consumer process
    or just a separate thread from the code that observes
    the things you want to measure.
    """
    consumer = BufferedConsumer()
    payload = queue.get()
    while payload is not None:
        consumer.send(*payload)
        payload = queue.get()

    consumer.flush()

if __name__ == '__main__':
    # replace token with your real project token
    token = '2118572b315ee841dd0f3e6fa9abdab3'
    distinct_id = ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789') for _ in range(32))

    queue = multiprocessing.Queue()
    sender = multiprocessing.Process(target=do_sending, args=(queue,))
    tracker = multiprocessing.Process(target=do_tracking,
                                      args=(token, distinct_id, queue))

    sender.start()
    tracker.start()
    tracker.join()
    sender.join()


mixpanel/consumer_tornado.py
__author__ = 'Fredrik Svensson'

import base64
import json
import urllib.parse
from tornado.httpclient import AsyncHTTPClient
import mixpanel


class AsyncConsumer(mixpanel.Consumer):
    """Sends requests directly using AsyncHTTPClient."""
    def _write_request(self, request_url, json_message):
        def handle_request(response):
            if response.error:
                raise mixpanel.MixpanelException(response.error)
            else:
                try:
                    response = json.loads(response.body.decode('utf-8'))
                except ValueError:
                    raise mixpanel.MixpanelException(
                        'Cannot interpret Mixpanel server response: {0}'.format(response)
                    )

                if response['status'] != 1:
                    raise mixpanel.MixpanelException(
                        'Mixpanel error: {0}'.format(response['error'])
                    )

        data = urllib.parse.urlencode({
            'data': base64.b64encode(json_message.encode('utf-8')),
            'verbose': 1,
            'ip': 0
        }).encode('utf-8')

        http_client = AsyncHTTPClient()
        http_client.fetch(request_url, handle_request, method='POST', body=data)


class BufferedAsyncConsumer(mixpanel.BufferedConsumer):
    """Same as BufferedConsumer but uses AsyncConsumer as base.

    TODO: How to flush? Give example code!
    """
    def __init__(self, max_size=50, events_url=None, people_url=None):
        self._consumer = AsyncConsumer(events_url, people_url)
        self._buffers = {
            'events': [],
            'people': []
        }
        self._max_size = min(50, max_size)


mixpanel/__init__.py
import base64
import json
import time
import urllib.request
import urllib.parse
import urllib.error

# The mixpanel package allows you to easily track events and
# update people properties from your python application.
#
# The Mixpanel class is the primary class for tracking events and
# sending people analytics updates.
#
# The Consumer and BufferedConsumer classes allow callers to
# customize the IO characteristics of their tracking.

VERSION = '3.0.4'


class Mixpanel(object):
    """
    Use instances of Mixpanel to track events and send Mixpanel
    profile updates from your python code.
    """

    def __init__(self, token, consumer=None):
        """
        Creates a new Mixpanel object, which can be used for all tracking.

        To use mixpanel, create a new Mixpanel object using your
        token.  Takes in a user token and an optional Consumer (or
        anything else with a send() method). If no consumer is
        provided, Mixpanel will use the default Consumer, which
        communicates one synchronous request for every message.
        """
        self._token = token
        self._consumer = consumer or Consumer()

    @staticmethod
    def _now():
        return time.time()

    def track(self, distinct_id, event_name, properties={}, meta={}):
        """
        Notes that an event has occurred, along with a distinct_id
        representing the source of that event (for example, a user id),
        an event name describing the event and a set of properties
        describing that event. Properties are provided as a Hash with
        string keys and strings, numbers or booleans as values.

          # Track that user "12345"'s credit card was declined
          mp.track("12345", "Credit Card Declined")

          # Properties describe the circumstances of the event,
          # or aspects of the source or user associated with the event
          mp.track("12345", "Welcome Email Sent", {
              'Email Template' => 'Pretty Pink Welcome',
              'User Sign-up Cohort' => 'July 2013'
          })
        """
        all_properties = {
            'token': self._token,
            'distinct_id': distinct_id,
            'time': int(self._now()),
            'mp_lib': 'python',
            '$lib_version': VERSION
        }
        all_properties.update(properties)
        event = {
            'event': event_name,
            'properties': all_properties
        }
        event.update(meta)
        self._consumer.send('events', json.dumps(event, separators=(',', ':')))

    def alias(self, alias_id, original, meta={}, consumer=None):
        """
        Gives custom alias to a people record.

        By default it will use the standard synchronous consumer, but you may
        provide another consumer if this is not wanted. Unlike other methods,
        this method will ignore any consumer object provided to the Mixpanel
        object on construction.

        Alias sends an update to our servers linking an existing distinct_id
        with a new id, so that events and profile updates associated with the
        new id will be associated with the existing user's profile and behavior.
        Example:
            mp.alias('amy@mixpanel.com', '13793')
        """

        sync_consumer = consumer or Consumer()
        event = {
            'event': '$create_alias',
            'properties': {
                'token': self._token,
                'distinct_id': original,
                'alias': alias_id
            }
        }
        event.update(meta)
        sync_consumer.send('events', json.dumps(event, separators=(',', ':')))
        # assert(False)

    def people_set(self, distinct_id, properties, meta={}):
        """
        Set properties of a people record.

        Sets properties of a people record given in JSON object. If the profile
        does not exist, creates new profile with these properties.
        Example:
            mp.people_set('12345', {'Address': '1313 Mockingbird Lane',
                                    'Birthday': '1948-01-01'})
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$set': properties
        }, meta=meta)

    def people_set_once(self, distinct_id, properties, meta={}):
        """
        Set immutable properties of a people record.

        Sets properties of a people record given in JSON object. If the profile
        does not exist, creates new profile with these properties. Does not
        overwrite existing property values.
        Example:
            mp.people_set_once('12345', {'First Login': "2013-04-01T13:20:00"})
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$set_once': properties
        }, meta=meta)

    def people_increment(self, distinct_id, properties, meta={}):
        """
        Increments/decrements numerical properties of people record.

        Takes in JSON object with keys and numerical values. Adds numerical
        values to current property of profile. If property doesn't exist adds
        value to zero. Takes in negative values for subtraction.
        Example:
            mp.people_increment('12345', {'Coins Gathered': 12})
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$add': properties
        }, meta=meta)

    def people_append(self, distinct_id, properties, meta={}):
        """
        Appends to the list associated with a property.

        Takes a JSON object containing keys and values, and appends each to a
        list associated with the corresponding property name. $appending to a
        property that doesn't exist will result in assigning a list with one
        element to that property.
        Example:
            mp.people_append('12345', {"Power Ups": "Bubble Lead"})
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$append': properties
        }, meta=meta)

    def people_union(self, distinct_id, properties, meta={}):
        """
        Merges the values for a list associated with a property.

        Takes a JSON object containing keys and list values. The list values in
        the request are merged with the existing list on the user profile,
        ignoring duplicate list values.
        Example:
            mp.people_union('12345', {"Items purchased": ["socks", "shirts"]})
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$union': properties
        }, meta=meta)

    def people_unset(self, distinct_id, properties, meta={}):
        """
        Removes properties from a profile.

        Takes a JSON list of string property names, and permanently removes the
        properties and their values from a profile.
        Example:
            mp.people_unset('12345', ["Days Overdue"])
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$unset': properties
        }, meta=meta)

    def people_delete(self, distinct_id, meta={}):
        """
        Permanently deletes a profile.

        Permanently delete the profile from Mixpanel, along with all of its
        properties.
        Example:
            mp.people_delete('12345')
        """
        return self.people_update({
            '$distinct_id': distinct_id,
            '$delete': ""
        }, meta=meta)

    def people_track_charge(self, distinct_id, amount, properties={}, meta={}):
        """
        Tracks a charge to a user.

        Record that you have charged the current user a certain amount of
        money. Charges recorded with track_charge will appear in the Mixpanel
        revenue report.
        Example:
            #tracks a charge of $50 to user '1234'
            mp.people_track_charge('1234', 50)

            #tracks a charge of $50 to user '1234' at a specific time
            mp.people_track_charge('1234', 50, {'$time': "2013-04-01T09:02:00"})
        """
        properties.update({'$amount': amount})
        return self.people_append(distinct_id, {'$transactions': properties},
                                  meta=meta)

    def people_clear_charges(self, distinct_id, meta={}):
        """
        Clears all charges from a user.

        Clears all charges associated with a user profile on Mixpanel.
        Example:
            #clear all charges from user '1234'
            mp.people_clear_charges('1234')
        """
        return self.people_unset(distinct_id, ["$transactions"], meta=meta)

    def people_update(self, message, meta={}):
        """
        Send a generic update to Mixpanel people analytics.
        Caller is responsible for formatting the update message, as
        documented in the Mixpanel HTTP specification, and passing
        the message as a dict to update. This
        method might be useful if you want to use very new
        or experimental features of people analytics from python
        The Mixpanel HTTP tracking API is documented at
        https://mixpanel.com/help/reference/http
        """
        record = {
            '$token': self._token,
            '$time': int(self._now() * 1000)
        }
        record.update(message)
        record.update(meta)
        self._consumer.send('people', json.dumps(record, separators=(',', ':')))

    def get_tracking_url(self, event_name, distinct_id=None, redirect=None,
                         image=False, properties={}, meta={}):
        """
        Returns a URL suitable to detect email open or click events.

        It's an error to set both image and redirect.
        If image is True the URL will return a 1x1 transparent pixel image.
        If redirect is set to a URL the client will be directed there.
        If no distinct_id is given mixpanel will use the IP of the request.
        """
        assert redirect is None or image is False, \
            "Makes no sense to set both image and redirect."

        all_properties = {'token': self._token}
        if distinct_id is not None:
            all_properties['distinct_id'] = distinct_id
        all_properties.update(properties)
        event = {
            'event': event_name,
            'properties': all_properties
        }
        event.update(meta)
        data = json.dumps(event, separators=(',', ':')).encode('utf-8')
        data = {'data': base64.b64encode(data)}
        if 'distinct_id' not in all_properties.keys():
            data['ip'] = 1
        if image:
            data['img'] = 1
        elif redirect:
            data['redirect'] = redirect

        return self._consumer.endpoints['events'] + \
            '?' + urllib.parse.urlencode(data)


class MixpanelException(Exception):
    """
    MixpanelExceptions will be thrown if the server can't receive
    our events or updates for some reason- for example, if we can't
    connect to the Internet.
    """
    pass


class Consumer(object):
    """
    The simple consumer sends an HTTP request directly to the Mixpanel service,
    with one request for every call. This is the default consumer for Mixpanel
    objects- if you don't provide your own, you get one of these.
    """
    def __init__(self, events_url=None, people_url=None):
        self.endpoints = {
            'events': events_url or 'https://api.mixpanel.com/track',
            'people': people_url or 'https://api.mixpanel.com/engage'
        }

    def send(self, endpoint, json_message):
        """
        Record an event or a profile update. Send is the only method
        associated with consumers. Will raise an exception if the endpoint
        doesn't exist, if the server is unreachable or for some reason
        can't process the message.

        All you need to do to write your own consumer is to implement
        a send method of your own.

        :param endpoint: One of 'events' or 'people', the Mixpanel endpoint for
            sending the data
        :type endpoint: str (one of 'events' or 'people')
        :param json_message: A json message formatted for the endpoint.
        :type json_message: str
        :raises: MixpanelException
        """
        if endpoint in self.endpoints:
            self._write_request(self.endpoints[endpoint], json_message)
        else:
            raise MixpanelException(
                'No such endpoint "{0}". Valid endpoints are one of {1}'.format(list(self.endpoints.keys()))
            )

    @staticmethod
    def flush():
        """To have the same interface as the BufferedConsumer."""
        pass

    def _write_request(self, request_url, json_message):
        data = urllib.parse.urlencode({
            'data': base64.b64encode(json_message.encode('utf-8')),
            'verbose': 1,
            'ip': 0
        }).encode('utf-8')
        try:
            req = urllib.request.Request(url=request_url, data=data,
                                         method='POST')
            f = urllib.request.urlopen(req)
            response = f.read().decode('utf-8')
        except urllib.error.HTTPError as e:
            raise MixpanelException(e)

        try:
            response = json.loads(response)
        except ValueError:
            raise MixpanelException(
                'Cannot interpret Mixpanel server response: {0}'.format(response)
            )

        if response['status'] != 1:
            raise MixpanelException(
                'Mixpanel error: {0}'.format(response['error'])
            )

        return True


class BufferedConsumer(object):
    """
    BufferedConsumer works just like Consumer, but holds messages in
    memory and sends them in batches. This can save bandwidth and
    reduce the total amount of time required to post your events.

    Because BufferedConsumers hold events, you need to call flush()
    when you're sure you're done sending them. calls to flush() will
    send all remaining unsent events being held by the BufferedConsumer.
    """
    def __init__(self, max_size=50, events_url=None, people_url=None):
        self._consumer = Consumer(events_url, people_url)
        self._buffers = {
            'events': [],
            'people': []
        }
        self._max_size = min(50, max_size)

    def send(self, endpoint, json_message):
        """
        Record an event or a profile update. Calls to send() will store
        the given message in memory, and (when enough messages have been stored)
        may trigger a request to Mixpanel's servers.

        Calls to send() may throw an exception, but the exception may be
        associated with the message given in an earlier call. If this is the
        case, the resulting MixpanelException e will have members e.message
        and e.endpoint

        :param endpoint: One of 'events' or 'people', the Mixpanel endpoint
            for sending the data
        :type endpoint: str (one of 'events' or 'people')
        :param json_message: A json message formatted for the endpoint.
        :type json_message: str
        :raises: MixpanelException
        """
        if endpoint not in self._buffers:
            raise MixpanelException(
                'No such endpoint "{0}". Valid endpoints are one of {1}'.format(list(self._buffers.keys()))
            )

        buf = self._buffers[endpoint]
        buf.append(json_message)
        if len(buf) >= self._max_size:
            self._flush_endpoint(endpoint)

    def flush(self):
        """
        Send all remaining messages to Mixpanel. BufferedConsumers will
        flush automatically when you call send(), but you will need to call
        flush() when you are completely done using the consumer (for example,
        when your application exits) to ensure there are no messages remaining
        in memory.

        Calls to flush() may raise a MixpanelException if there is a problem
        communicating with the Mixpanel servers. In this case, the exception
        thrown will have a message property, containing the text of the message,
        and an endpoint property containing the endpoint that failed.

        :raises: MixpanelException
        """
        for endpoint in list(self._buffers.keys()):
            self._flush_endpoint(endpoint)

    def _flush_endpoint(self, endpoint):
        buf = self._buffers[endpoint]
        while buf:
            batch = buf[:self._max_size]
            batch_json = '[{0}]'.format(','.join(batch))
            try:
                self._consumer.send(endpoint, batch_json)
            except MixpanelException as e:
                e.message = 'batch_json'
                e.endpoint = endpoint
            buf = buf[self._max_size:]
        self._buffers[endpoint] = buf
